src/OpenTelemetry/Trace/ActivitySourceAdapter.cs

```diff
@@ -122,18 +126,213 @@ internal void UpdateProcessor(BaseProcessor<Activity> processor)
             return Expression.Lambda<Action<Activity, ActivityKind>>(body, instance, propertyValue).Compile();
         }
 
-        private void RunGetRequestedDataAlwaysOnSampler(Activity activity)
+        /*
+            Build a dynamic method for invoking the sample method on all listeners.
+
+            Based on this: https://github.com/dotnet/runtime/blob/f2148079d4476073bb5e79277b557807ed0c9984/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs#L194-L216
+
+            SampleResult Sample(ActivitySource activitySource, Activity activity, ActivityContext activityContext)
+            {
+                SynchronizedList<ActivityListener>? listeners = _listeners;
+                if (listeners == null || listeners.Count == 0)
+                {
+                    return new SampleResult(ActivitySamplingResult.None, null);
+                }
+
+                ActivitySamplingResult samplingResult = default;
+
+                ActivityCreationOptions<ActivityContext> aco = new ActivityCreationOptions<ActivityContext>(activitySource, activity.DisplayName, activityContext, activity.Kind, activity.TagObjects, activity.Links);
+
+                listeners.EnumWithFunc((ActivityListener listener, ref ActivityCreationOptions<ActivityContext> data, ref ActivitySamplingResult result, ref ActivityCreationOptions<ActivityContext> unused) => {
+                    SampleActivity<ActivityContext>? sample = listener.Sample;
+                    if (sample != null)
+                    {
+                        ActivitySamplingResult dr = sample(ref data);
+                        if (dr > result)
+                        {
+                            result = dr;
+                        }
+                    }
+                }, ref aco, ref samplingResult, ref aco);
+
+                return new SampleResult(samplingResult, aco._samplerTags);
+            }
+        */
+        private static Func<ActivitySource, Activity, ActivityContext, SampleResult> CreateCallSamplersFunc()
```

Comment:
@tarekgh This was mega hard to put together 🤯 We don't have to expose anything in the public API, but if you could refactor the below block into a helper method that I could bind to it would be much easier and probably less fragile:

https://github.com/dotnet/runtime/blob/f2148079d4476073bb5e79277b557807ed0c9984/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs#L195-L206

The main challenge with it as-is (primarily) is the closure. Have to hook into compiler-generated types to fire that.